Operator modular metrics

pippolo84 opened the pull request:
Add a cell to provide global metrics registry and prometheus metrics server for the operator hive.

This makes possible to use `cell.Metric` to export metrics in the operator cells.

Since there are parts of the operator code not yet modularized, the legacy metrics variables and the global registry are kept exported, in order to be directly used by legacy code. This approach is inspired by the agent metrics cell.

Besides, this PR fixes the operator metrics namespace, setting it early at startup, so that all the code shared between other components (e.g: the agent) will report the correct `cilium_operator` namespace when updating the metrics.

Here a query that shows both the correct namespace for all the metrics and also all the metrics registered with `cell.Metric` (like the ones in the `operator/pkg/lbipam` cell) showing up:

```
$ curl localhost:9234/v1/metrics
[
    {
        "name": "cilium_operator_ces_queueing_delay_seconds"
    },
    {
        "name": "cilium_operator_ces_sync_errors_total"
    },
    {
        "name": "cilium_operator_lbipam_conflicting_pools_total"
    },
    {
        "name": "cilium_operator_lbipam_services_matching_total"
    },
    {
        "name": "cilium_operator_lbipam_services_unsatisfied_total"
    },
    {
        "name": "cilium_operator_number_of_ceps_per_ces"
    },
    {
        "name": "cilium_operator_process_cpu_seconds_total",
        "value": 1.03
    },
    {
        "name": "cilium_operator_process_max_fds",
        "value": 1048576
    },
    {
        "name": "cilium_operator_process_open_fds",
        "value": 12
    },
    {
        "name": "cilium_operator_process_resident_memory_bytes",
        "value": 95997952
    },
    {
        "name": "cilium_operator_process_start_time_seconds",
        "value": 1694106767.7
    },
    {
        "name": "cilium_operator_process_virtual_memory_bytes",
        "value": 1351573504
    },
    {
        "name": "cilium_operator_process_virtual_memory_max_bytes",
        "value": 18446744073709552000
    }
]
```

pippolo84 committed:
operator: Set metrics namespace before starting

Each cell that adds metrics through cell.Metric refers the global
variable metrics.Namespace.  This variable is set at startup with the
value "cilium", that is, the agent metrics namespace.  For cells that
are also part of the operator hive, like hive/job, this is incorrect.

This commit overwrites the metrics.Namespace global variable with the
correct one for the operator before running the hive.

Doing this a separate global variable for the operator namespace is not
needed anymore, so the unmodularized operator code is updated to refer
to the global metrics.Namespace.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator: Modularize metrics

Add a cell to provide global metrics registry and prometheus metrics
server for the operator hive.

This makes possible to use cell.Metric to export metrics in the operator
cells.

Since there are parts of the operator code not yet modularized, the
legacy metrics variables and the global registry are kept exported, in
order to be directly used by legacy code.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator: Remove check for EnableMetrics flag

Operator legacy metrics have been changed to be declared as
metrics.NoOp*. Thus, even when the metrics are not enabled in the
operator, all the usual methods can be safely called on them. Therefore,
there is no need to check the EnableMetrics option every time a metric
is referred in the operator code.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on operator/metrics/cell.go:
```diff
@@ -0,0 +1,59 @@
+	flags.String(OperatorPrometheusServeAddr, def.OperatorPrometheusServeAddr, "Address to serve Prometheus metrics")
+}
+
+type MetricsConfig struct {
```
This can be removed?

pippolo84 replied:
Yep, that's just a leftover from a previous iteration :sweat_smile:

joamaki commented on operator/metrics/legacy.go:
```diff
@@ -0,0 +1,212 @@
+)
+
+type legacyMetrics struct {
+	IdentityGCSize                metric.Vec[metric.Gauge]
```
Let's do a follow-up to remove `legacyMetrics`. Should be pretty doable to do all these after the CES PR lands?

pippolo84 replied:
Definitely. Both the identity GC and the CES controller have been modularized, so we can switch to `cell.Metric` and remove all of this. :+1:

derailed commented on operator/metrics/legacy.go:
```diff
@@ -0,0 +1,212 @@
+			Namespace: metrics.CiliumOperatorNamespace,
+			Name:      "ces_sync_total",
+			Help:      "The number of completed CES syncs by outcome",
+		}, []string{"outcome"}),
```
Should we use the defined label aka LabelOutcome?

derailed commented on operator/metrics/legacy.go:
```diff
@@ -0,0 +1,212 @@
+		metricType := val.GetType()
+
+		for _, metricLabel := range val.Metric {
+			labels := map[string]string{}
```
nit: prealloc the map? ie make(map[string]string, len(metricLabel.GetLabel()))

derailed reviewed (commented):
@pippolo84 Nice work Fabio! Just a couple comments...

pippolo84 committed:
operator/identitygc: Remove useless enableMetrics bool

Operator metrics have been switched to metrics.NoOp* types, so there is
no need to keep each metrics operation behind a flag. The enableMetrics
field is a leftover from the previous metrics management and thus should
be removed.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

derailed reviewed (approved):
@pippolo84 LGTM. Thanks for the updates!

tommyp1ckles reviewed (approved):
IPAM changes look good
